File: leetcode/golang/4.go
package leetcode

func getUpperBound(nums []int, target int) int {
	l, r, ret := 0, len(nums)-1, len(nums)
	for l <= r {
		mid := (r-l)/2 + l
		if nums[mid] > target {
			ret = mid
			r = mid - 1
		} else {
			l = mid + 1
		}
	}
	return ret
}

func findKthNumber(k int, nums1 []int, nums2 []int) int {
	if len(nums1) == 0 {
		return nums2[k-1]
	}
	if len(nums2) == 0 {
		return nums1[k-1]
	}
	mid1 := (len(nums1) - 1) / 2
	idx2 := getUpperBound(nums2, nums1[mid1])
	cnt1, cnt2 := mid1, idx2
	if k == cnt1+cnt2+1 {
		return nums1[mid1]
	} else if k <= cnt1+cnt2 {
		return findKthNumber(k, nums1[0:mid1], nums2[0:idx2])
	} else {
		return findKthNumber(k-cnt1-cnt2-1, nums1[mid1+1:], nums2[idx2:])
	}
}

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	total := len(nums1) + len(nums2)

	if total%2 == 0 {
		m1 := findKthNumber(total/2, nums1, nums2)
		m2 := findKthNumber(total/2+1, nums1, nums2)
		return 0.5 * float64(m1+m2)
	}

	return float64(findKthNumber(total/2+1, nums1, nums2))
}

File: leetcode/cpp/4.cpp
#include "leetcode.h"

class Solution {
private:
	int findKthNumber(int k, const vector<int> &nums1, const int s1, const int e1,
						const vector<int> &nums2, const int s2, const int e2) // [s, e), don't do out of access check
	{
		if (e1 <= s1) {
			return nums2[s2+k-1];
		}
		if (e2 <= s2) {
			return nums1[s1+k-1];
		}
		int mid1 = (e1 - s1 - 1) / 2 + s1;
		int lcnt1 = mid1 - s1 + 1;
		int mid1Value = nums1[mid1];
		int idx2 = std::upper_bound(nums2.begin() + s2, nums2.begin() + e2, mid1Value) 
					- nums2.begin() - 1;
		int lcnt2 = idx2 - s2 + 1;
		if (lcnt1 + lcnt2 == k) {
			return mid1Value;
		} else if (lcnt1 + lcnt2 > k) {
			return findKthNumber(k, nums1, s1, mid1, nums2, s2, idx2+1);
		} else { // (lcnt1 + lcnt2 < k) 
			return findKthNumber(k-lcnt1-lcnt2, nums1, mid1+1, e1, nums2, idx2+1, e2);
		}
	}

public:
	double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2)
	{
		int total = nums1.size() + nums2.size();
		if (total%2 == 0) {
			int m1 = findKthNumber(total/2, nums1, 0, nums1.size(), nums2, 0, nums2.size());
			int m2 = findKthNumber(total/2+1, nums1, 0, nums1.size(), nums2, 0, nums2.size());
			return 0.5 * (m1+m2);
		} else {
			return findKthNumber(total/2+1, nums1, 0, nums1.size(), nums2, 0, nums2.size());
		}
	}
};